### PR TITLE
DRAFT JAMES-4824 Refactor quota for Cassandra: create POJOs and DAOs

### PR DESCRIPTION
**Pull request by hungphan227**

### CODE REVIEW

File: core/src/main/java/org/apache/james/core/quota/QuotaCurrentValue.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+package org.apache.james.core.quota;
+
+import java.util.Objects;
+
+public class QuotaCurrentValue {
+
+    private final String identifier;
+    private final QuotaComponent quotaComponent;
+    private final QuotaType quotaType;
+    private final Long currentValue;
```
Comment (chibenwa) [inline core/src/main/java/org/apache/james/core/quota/QuotaCurrentValue.java:29]: Why a `Long`? Are you using `null` so that you need a boxed value? But if so nullity is truely dangerous and should be avoided...

Alternative (more explicit): `Optional<Long>`

File: core/src/main/java/org/apache/james/core/quota/QuotaLimit.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+import java.util.Objects;
+
+public class QuotaLimit {
+
+    private final String identifier;
+    private final QuotaComponent quotaComponent;
+    private final QuotaType quotaType;
+    private final QuotaScope quotaScope;
+    private final Long maxValue;
```
Comment (chibenwa) [inline core/src/main/java/org/apache/james/core/quota/QuotaLimit.java:30]: Idem

File: core/src/main/java/org/apache/james/core/quota/QuotaLimit.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    private QuotaLimit(String identifier, QuotaComponent quotaComponent, QuotaType quotaType, QuotaScope quotaScope, Long maxValue) {
+        this.identifier = identifier;
+        this.quotaComponent = quotaComponent;
+        this.quotaType = quotaType;
+        this.quotaScope = quotaScope;
+        this.maxValue = maxValue;
+    }
+
+    public static QuotaLimit of(String identifier, QuotaComponent component, QuotaType quotaType, QuotaScope scope, Long maxValue) {
```
Comment (chibenwa) [inline core/src/main/java/org/apache/james/core/quota/QuotaLimit.java:40]: Static methods goes first by project convention

File: core/src/main/java/org/apache/james/core/quota/QuotaModelModule.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+import org.apache.james.core.quota.quotascope.GlobalQuotaScope;
+import org.apache.james.core.quota.quotascope.UserQuotaScope;
+import org.apache.james.core.quota.quotatype.CountQuotaType;
+import org.apache.james.core.quota.quotatype.SizeQuotaType;
+
+import com.google.inject.AbstractModule;
+import com.google.inject.Scopes;
+import com.google.inject.multibindings.Multibinder;
+
+public class QuotaModelModule extends AbstractModule {
```
Comment (chibenwa) [inline core/src/main/java/org/apache/james/core/quota/QuotaModelModule.java:35]: This has NOTHING to do in `/core`. If needed move this to `/server/container/guice`

File: core/src/main/java/org/apache/james/core/quota/QuotaScope.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.core.quota;
+
+public interface QuotaScope {
```
Comment (chibenwa) [inline core/src/main/java/org/apache/james/core/quota/QuotaScope.java:22]: Idem I think a simple value class wrapping a String would serve us better here.

Simpler is better.

File: core/src/main/java/org/apache/james/core/quota/QuotaScopeFactory.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.core.quota;
+
+import java.util.Set;
+
+import com.google.inject.Inject;
+
+public class QuotaScopeFactory {
```
Comment (chibenwa) [inline core/src/main/java/org/apache/james/core/quota/QuotaScopeFactory.java:26]: My first reflex would be to kill the factory

File: core/src/main/java/org/apache/james/core/quota/QuotaType.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.core.quota;
+
+public interface QuotaType {
```
Comment (chibenwa) [inline core/src/main/java/org/apache/james/core/quota/QuotaType.java:22]: Idem

File: core/src/main/java/org/apache/james/core/quota/QuotaTypeFactory.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.core.quota;
+
+import java.util.Set;
+
+import com.google.inject.Inject;
+
+public class QuotaTypeFactory {
```
Comment (chibenwa) [inline core/src/main/java/org/apache/james/core/quota/QuotaTypeFactory.java:26]: Idem

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/modules/CassandraQuotaModule.java [hunk 9]
```diff
@@ -68,5 +70,26 @@ public interface CassandraQuotaModule {
             .withPartitionKey(CassandraGlobalMaxQuota.KEY, TEXT)
             .withColumn(CassandraGlobalMaxQuota.STORAGE, BIGINT)
             .withColumn(CassandraGlobalMaxQuota.MESSAGE, BIGINT))
+
+        .table(CassandraQuotaLimit.TABLE_NAME)
+        .comment("Holds quota limit.")
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/modules/CassandraQuotaModule.java:75]:
```suggestion
        .comment("Holds quota limits.")
```

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/modules/CassandraQuotaModule.java [hunk 10]
```diff
@@ -68,5 +70,26 @@ public interface CassandraQuotaModule {
...
+
+        .table(CassandraQuotaCurrentValue.TABLE_NAME)
+        .comment("Holds quota limit.")
+        .options(options -> options
+            .withCaching(true, rows(CassandraConstants.DEFAULT_CACHED_ROW_PER_PARTITION)))
+        .statement(statement -> types -> statement
+            .withPartitionKey(CassandraQuotaCurrentValue.IDENTIFIER, TEXT)
+            .withPartitionKey(CassandraQuotaCurrentValue.QUOTA_COMPONENT, TEXT)
+            .withPartitionKey(CassandraQuotaCurrentValue.QUOTA_TYPE, TEXT)
+            .withColumn(CassandraQuotaCurrentValue.CURRENT_VALUE, BIGINT))
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/modules/CassandraQuotaModule.java:93]: Use a counter!

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraQuotaCurrentValueDao.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+        this.setQuotaCurrentValueStatement = session.prepare(setQuotaCurrentValueStatement().build());
+        this.quotaComponentFactory = quotaComponentFactory;
+        this.quotaTypeFactory = quotaTypeFactory;
+    }
+
+    Mono<QuotaCurrentValue> getQuotaCurrentValue(String identifier, QuotaComponent quotaComponent, QuotaType quotaType) {
+        return queryExecutor.executeSingleRow(getQuotaCurrentValueStatement.bind(identifier, quotaComponent.asString(), quotaType.asString()))
+            .map(row -> QuotaCurrentValue.of(row.get(CassandraQuotaCurrentValue.IDENTIFIER, String.class), quotaComponentFactory.parse(row.get(CassandraQuotaCurrentValue.QUOTA_COMPONENT, String.class)),
+                quotaTypeFactory.parse(row.get(CassandraQuotaCurrentValue.QUOTA_TYPE, String.class)), row.get(CassandraQuotaCurrentValue.CURRENT_VALUE, Long.class)));
+    }
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraQuotaCurrentValueDao.java:65]: Not readable a first site (6am here)

Please try extractoring IE row -> POJO convertion

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraQuotaCurrentValueDao.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+
+    Mono<Void> setQuotaCurrentValue(QuotaCurrentValue quotaCurrentValue) {
+        return queryExecutor.executeVoid(setQuotaCurrentValueStatement.bind(quotaCurrentValue.getIdentifier(),
+            quotaCurrentValue.getQuotaComponent().asString(), quotaCurrentValue.getQuotaType().asString(), quotaCurrentValue.getCurrentValue()));
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraQuotaCurrentValueDao.java:69]: Use named values in your binds

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraQuotaLimitDao.java [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+        this.quotaTypeFactory = quotaTypeFactory;
+        this.quotaScopeFactory = quotaScopeFactory;
+    }
+
+    Mono<QuotaLimit> getQuotaLimit(String identifier, QuotaComponent quotaComponent, QuotaType quotaType, QuotaScope quotaScope) {
+        return queryExecutor.executeSingleRow(getQuotaLimitStatement.bind(identifier, quotaComponent.asString(), quotaType.asString(), quotaScope.asString()))
+            .map(row -> QuotaLimit.of(row.get(CassandraQuotaLimit.IDENTIFIER, String.class), quotaComponentFactory.parse(row.get(CassandraQuotaLimit.QUOTA_COMPONENT, String.class)),
+                quotaTypeFactory.parse(row.get(CassandraQuotaLimit.QUOTA_TYPE, String.class)), quotaScopeFactory.parse(row.get(CassandraQuotaLimit.QUOTA_SCOPE, String.class)),
+                    row.get(CassandraQuotaLimit.MAX_VALUE, Long.class)));
+    }
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraQuotaLimitDao.java:70]: Idem method extraction please

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraQuotaLimitDao.java [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+
+    Mono<Void> setQuotaLimit(QuotaLimit quotaLimit) {
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraQuotaLimitDao.java:72]: Increase / decrease quta limit in order to fit the cassandra counter semantic.

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/quota/CassandraQuotaCurrentValueDaoTest.java [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+import org.apache.james.core.quota.quotacomponent.MailBoxQuotaComponent;
+import org.apache.james.core.quota.quotatype.SizeQuotaType;
+import org.apache.james.mailbox.cassandra.mail.utils.GuiceUtils;
+import org.apache.james.mailbox.cassandra.modules.CassandraQuotaModule;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+public class CassandraQuotaCurrentValueDaoTest {
```
Comment (chibenwa) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/quota/CassandraQuotaCurrentValueDaoTest.java:35]: Please better test edge cases...

  - set A, set B
  - Negative values
  - 0
  - etc...

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraQuotaCurrentValueDao.java [hunk 16]
```diff
@@ -0,0 +1,88 @@
+/****************************************************************
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraQuotaCurrentValueDao.java:1]: No way to remove a current value in this DAO?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraQuotaLimitDao.java [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailbox.cassandra.quota;
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraQuotaLimitDao.java:20]: No way to remove a limit in this DAO?

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/quota/CassandraQuotaLimitDaoTest.java [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+import org.apache.james.core.quota.quotascope.UserQuotaScope;
+import org.apache.james.core.quota.quotatype.SizeQuotaType;
+import org.apache.james.mailbox.cassandra.mail.utils.GuiceUtils;
+import org.apache.james.mailbox.cassandra.modules.CassandraQuotaModule;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+public class CassandraQuotaLimitDaoTest {
```
Comment (chibenwa) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/quota/CassandraQuotaLimitDaoTest.java:36]: Better test your edge cases

**Review by chibenwa — COMMENTED** [review]
IMO a good indicator here would be to cut ~halth the count of lines of code.

**Comment by chibenwa** [comment]
I am not sure a memory implementation is needed as so far the quota helpers are scoped to cassandra...
That 's a cassandra implementation detail IMO.

**Comment by Arsnael** [comment]
Alright then, two tasks POJOs and DAOs

File: core/src/main/java/org/apache/james/core/quota/QuotaComponentFactory.java [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.core.quota;
+
+import java.util.Set;
+
+import com.google.inject.Inject;
+
+public class QuotaComponentFactory {
```
[the remark this answers is not in the document]
Comment (Arsnael) [reply core/src/main/java/org/apache/james/core/quota/QuotaComponentFactory.java:26]: @chibenwa : I think @hungphan227 is concerned at the end that with this simple way, we might not be able to tell the user if we have quotas for the component or not when calling the webadmin api generic route. 

Maybe you don't care if we return an empty list in that case? Or maybe you don't want generic routes in the first place, and want to keep routes separated like it is the case now? (I think would be easier to have a generic route with the quota component in it though, as groomed)
